README.md
# SolvisSmartHomeServer
## Überblick

Dieses Programnm dient der Anbindung der SolvisMax 6 und 7 mit SolvisControl2 in Kombination der SolvisRemote an Smart-Home-Systeme wie z.B. Fhem, ioBroker, OpenHAB, Indigo.

Es eignet sich nicht für die neueren Anlagen der SolvisMax7 und SolvisBen, welche mit der neuen SolvisControl3 ausgeliefert werden. Bei diesen sollte man zur Smart-Home-System-Anbindung das Modbus-Interface nutzen.

Dieses Programm ist ein eigenständiges Java-Programm, das als Service/Task/Daemon im Hintergrund läuft. Ziel war nicht nur ein Monitoring der Anlage zu ermöglichen sondern auch die Einstellung der wichtigsten Anlageparameter wie z. B. Soll-Temperaturen, Raumabhängigkeit, Anlagemodus.

Die bisher mir bekannten Lösungen liefern nur die Messwerte/Zustände einer Solvis-Anlagen an die SmartHome-Systeme. Dabei wird ein XML-String interpretiert,
der über das WebInterface der SolvisRemote abgefragt wird. Der SolvisSmartHomeServer nutzt ebenfalls diese Schnittstelle.
Eine Steuerung der Anlage mit dieser Schnittstelle ist jedoch nicht möglich.

Erst bei neueren Anlagen (von der Solvis-Control-2-Zentralreglerversion MA205 an) hat die Firma Solvis die Steuermöglichkeit über das Modbus-Interface realisiert. Ältere Anlagen bleiben dabei außen vor.

Bei älteren Anlagen gibt es nur die Möglichkeit der Steuerung über die Web-Oberfläche der SolvisRemote, das jedoch auf rein grafischer Basis arbeitet
(es ist eine Pixelkopie der SolvisControl2 der Anlage). Die Steuerung darüber erfolgt über Maus-Klicks auf bestimmte Koordinaten auf diesem GUI,
die Rückmeldung erfolgt ebenfalls nur grafisch über das GUI.

### Funktionsweise der Einstellung der Anlageparameter

Dieses Projekt nutzt diesen Weg. Bei einer Sollwert-Änderung wird mit Hilfe eines OCRs zunächst der aktuelle Wert grafisch aus der Web-Oberfläche ermittelt. Anschließend wird der Wert entsprechend der Anforderung durch entsprechende simulierte Maus-Klicks geändert. Zum Abschluss wird diese Änderung durch das OCR verifiziert.

![Gui](https://raw.githubusercontent.com/GollmerSt/SolvisSmartHomeServer/master/docu/images/Hierarchie/1%20Heizung.png)

So wird bei einer Änderung der Feineinstellung die 0 in den eckigen Klammern im oben angezeigten "Home-Screen" mittels OCR erkannt und abhängig
vom einzustellenden Wert die Plus/Minus-Buttons auf der rechten Seite solange betätigt, bis der richtige Wert zwischen den eckigen Klammern
erscheint. Ist der zu ändernde Wert nicht auf dem "Home-Screen", so wird automatisch der entsprechende Screen angefahren. Das erfolgt z.B.
durch die Buttons auf der linken Seite.

### Features
* Auslesen der Messwerte der Sensoren
* Einstellung der Anlagenparameter wie Temperatur-Sollwerte, Raumabhängigkeiten etc.
* Monitoring der Solvis-Uhr mit entsprechender Nachjustierung.
* Erkennen eines Fehlerzustandes der Anlage. Im Fehlerfall kann optional eine Mail versendet werden.
Wird der Fehler als Fehlerscreen von der Anlage angezeigt, wird an die Mail die Hardcopy des Bildschirmes angehängt.
Hier ein Beispiel einer solchen Fehlermeldung:

   ![Fehlermeldung](https://raw.githubusercontent.com/GollmerSt/SolvisSmartHomeServer/master/testFiles/images/Stoerung%205.png)
* Es werden Anwender und Service-Zugriffe auf den Touchscreen der SolvisControl erkannt und nach beenden die möglicherweise
veränderten Anlagenparameter wieder erneut gelesen.
* Anbindung über MQTT, damit ist das System in SmartHome-Systeme ohne speziellen Client möglich.
Voraussetzung ist nur eine MQTT-Schnittstelle(-Erweiterung) im SmartHome-System sowie ein MQTT-Broker (z.B. Mosquitto). U.U.
bringt ein Smart-Home-System einen solchen Broker das selber mit.
* Zusätzliche Möglichkeit der Anbindung über eine Client-Server-Verbindung
  * Daten zwischen Server-Client werden im JSON-Format ausgetauscht
  * Es können sich max. 50 Clients mit dem Server verbinden
* Leichte Anpassungsmöglichkeit an vorhandene Anlage über XML-Files. Die XML Schema sind mit enthalten, so dass Anpassung
mittels XML-Editor (z.B. integriert in Eclipse) stark vereinfacht wird. Ab Version 1.03.00 sind für SolvisMax 6 und 7 die
laut der Anlagenschema-Dokumentation von Solvis möglichen Konfigurationen berücksichtigt.

Eine Liste der Kanäle findet sich u.a. im Kapitel [MQTT-Schnittstelle](https://github.com/GollmerSt/SolvisSmartHomeServer/wiki/MQTT-Schnittstelle#aktuell-definierte-channels) bzw. kann ab Version 1.03.00 vom Programm selber passend zur Anlagenkonfiguartion generiert werden. Dies ist [hier](https://github.com/GollmerSt/SolvisSmartHomeServer/wiki/Dokumentation-der-Kanalbelegung) beschrieben.

### Voraussetzungen
* Solvis-Heizungsanlage (SolvisMax6/7) mit einer SolvisControl **2** und SolvisRemote. Bei einer SolvisControl **3** ist der
Einsatz nicht möglich, da das Web-Interface durch das Solvis-Portal ersetzt wurde. Zur Anbindung an ein Smart-Home-System
sollte man dort den Modbus verwenden.
* PC/Raspberry o.ä. mit Linux oder Windows, mit JRE >= 1.8.
* Ein Smart-Home-System wie FHEM, ioBroker, OpenHab o.ä. mit MQTT Schnittstelle.

### Interfaces zu Smart-Home-Systemen
Die Anbindung an das SmartHome-System werden durch zwei verschiedene Interfaces ermöglicht:

1. MQTT-Schnittstelle
1. Eine proprietäre Server-Client-Schnittstelle. Diese Schnittstelle erfordert auf der Smart-Home-System-Seite einen speziellen Client.
Ein solcher existiert bisher nur für das FHEM-SmartHome-System

Vom Funktionsumfang her gibt es zwischen beiden Interfaces keine Unterschiede. Anfangs gab es nur die proprietäre Server-Client-Schnittstelle.
Um bei SmartHome-Systemen nicht auf ein spezielles Modul angewiesen zu sein, wurde die MQTT-Schnittstelle zusätzlich implementiert. Für die
meisten SmartHome-System gibt es für die MQTT-Schnittstelle entsprechende Module.

#### Es sind für folgende SmartHome-Systeme Anpassungen vorhanden:
Smart-Home-System | Beschreibung
------------------ | ------------
FHEM | Modul, das auf der Server-Client-Schnittstelle basiert (Teil des Progarmmpaketes).
ioBroker | Objektliste mit Pairing-Script für das ioBroker-Modul MQTT Client (Wird vom Programm konfigurationsspezifisch generiert).

Weitere sind in Vorbereitung

### Installation
Die genauen Installationsanweisungen enthält die unten angehängte ausführliche Dokumentation. Hier daher nur ein grober Überblick.

Das Release beinhaltet normalerweise immer 2 unterschiedliche Installationspakete:
* Installationspaket für Linux mit einem Makefile, das der Installation dient
* Installationspaket für Windows mit einem Installer

**Vor** der Ausführung der Installationspakete ist es wichtig, dass bereits Java installiert ist, da die Installation selber schon Java-Programme
ausführen. Ich verwende unter Windows die Oracle-Java-Version 1.8, oder die OpenJDK ab 8, aktuell 14. Unter Linux habe ich bisher die Version
ab JDK 8 verwendet, vom Funktionsumfang müsste noch die Version 7 gehen.

Der Ablauf der Installation sollte folgendermaßen ablaufen:

1. Installation von Java
1. Herunterladen und Auspacken des Zip-File-Release in einem Ordner
1. Unter Windows Aufruf des Installers
1. Kopieren des base.xml.new unter dem Namen base.xml
1. Anpassen des base.xml (Solvis-IP, Passwörter, etc.), in einer zukünftigen Version werde ich hier einen Konfigurator noch zur Verfügung stellen
1. Aufruf des Learning (über Startmenü unter Windows oder über das Makefile unter Linux). Ein Learning ist notwendig, da die unterschiedlichen
Screens grafisch erkannt werden. Je nach eingestellter Sprache unterscheiden sich die Screens entsprechend. Auch die Erkennung der Symbole (Standby,
Nacht, Tag etc.) wird hierbei vom Screen gelesen und gemerkt.
1. Start des Servers

Die genaue Vorgehensweise muss man aus der folgenden Dokumentation entnommen werden:

## Ausführliche Dokumentation
Eine umfangreiche Dokumentation der Installation, Interfaces, Arbeitsweise etc. findet sich [in der Wiki](https://github.com/GollmerSt/SolvisSmartHomeServer/wiki).


docu/CHANGES.txt
A E N D E R U N G E N

Version 01.06.00

IControlAccess

    25.07.2022  Bei Human access wird der Typ des Screens als erstes verglichen (z.B. Warmwasser).

    24.07.2022  Das Anlagen-Fehlerhandling wurde komplett ueberarbeitet, da beim alten
                Fehlermeldungen verloren gingen, wenn ein Fehler waehrend einer
                Parameterveraenderung erfolgte. Ausserdem funktionierte die Fehlerbehandlung nicht
                immer korrekt, wenn das Feature "ClearErrorMessageAfterMail" nicht aktiviert war.

    23.06.2022  Das Human-Access-Handling wurde in der Klasse HumanAccess zusammengefasst

    21.06.2022  Der Image-Vergleich wurde verschnellert.

    19.06.2022  Das Interface "IControlAccess" wurde ausgebaut, da der Server die Modus-
                Schnittstelle nicht bedient, da der Server bei Anlagen mit Modbus keinen
                Sinn macht.

    26.05.2022  Wird ein Mail-Test ausgefuehrt, werden fuer die Units eine Fehlermeldung
                ausgegeben, bei denen das Feature "SendMailOnError" nicht gesetzt ist. Der Test
                findet davon unabhaengig statt.
                
                Feature "SendMailOnErrorsCleared" hinzugefuegt. Ist dieses Feature auf true oder
                nicht definiert, so wird nach einer FehlerMail auch eine Mail versendet, wenn der
                Fehler behoben ist. Andernfalls wird nur der Fehler selber gesendet.

    19.04.2022  C49.Puffer_Tmin hinzugefuegt.

    15.04.2022  Vereinfachung der Nachbearbeitung der XML-Daten durch Ausnutzen des
                Nachprozessierens der XML-Library.

    14.04.2022  Ist in den Features der Admin-Mode gesetzt, so werden nun die Werte
                C47.Puffer_dT_Start und C48.Puffer_dT_Stopp ausgelesen. Sie koennen auch veraendert
                werden.

    07.02.2022  In Zukunft muss ich aus der Ferne mit unterschiedlichen IPs die Solvis-Anlage
                ansprechen koennen. Aus diesem Grund habe ich ein auf Git undokumentiertes XML-Tag
                Urls in den Unit-Tag der base.xml eingefuehrt. In dem entsprechenden xsd-File ist
                der Tag dokumentiert. Auf diese Weise koennen unterschioedliche Urls fuer die
                Anlage definiert werden.


Version 01.05.01

    27.12.2021  Die CLI-Option "--server-terminate" funktionierte nicht mehr (seit V01.05.01-rc1).

    27.12.2021  Es werden MailExceptionTag-Fehler nur dann ausgeben, wenn diese auch aktiviert ist.

    27.12.2021  Das Makefile ueberarbeitet um die FHEM-Abhaengigkeit zu beseitigen. So wird ein
                eigener User "solvis" erzeugt, unter dem der Server laeuft. Auch das Default-
                Verzeichnis in der base.xml wurde entsprechend auf "/opt/solvis" abgeaendert.

    21.12.2021  Die Aenderung des rc5 war noch etwas Quick & Dirty. Der Wert "maxPixelsOfEmptyLine"
                wurde daher in der control.xml eingefuehrt.


Version 01.05.01-rc5

    20.12.2021  Mit dem Support fuer verschiedene Sprachen mussten die Scanbereiche nach rechts
                erweitert werden. Bei einem groesseren Verschiebung durch die Solvis-Bildschirm-
                kalibrierung geriet dabei jedoch der obere Rand eines Buttons in den Scan-Bereich.
                Um dieses Problem zu umgehen, wurde eine Erkennung der vertikalen Lage der Bild-
                schirmtitel implementiert. Dadurch werden vertikale Offsets der Bildschirm-
                kalibrierung kompensiert.


Version 01.05.01-rc4

    17.12.2021  Bei einem Schreiben erfolgt nun eine Warnung, wenn der Schreibwert null ist.
                Das erleichtert die Fehlersuche bei soche Fehler wie vom 16.12.2021. 
                
                Rundungsfehler bei negativen Set-Werten behoben.


Version 01.05.01-rc3

    16.12.2021  Bei Control Read wurde nicht beruecksichtigt, dass der Set-Wert auch ein Long sein
                kann.


Version 01.05.01-rc2

    07.12.2021  Im Fall einer Brenner-Laufzeit-Synchronisation bleibt Control auf Monitoring, auch
                die Synchronisation abgeschlossen war. Gefixt.
                
    15.12.2021  JSON-Package-Exception-Handling verbessert 

    16.12.2021  TinyLog aktualisiert auf 2.4.1 


Version 01.05.01-rc1

    02.12.2021  Wird uwber die JSON-Schnittstelle ein SET-Befehl empfangen, die dem der Daten-
                Parameter leer ist oder nicht zu interpretieren ist, erfolgt faelschlicherweise eine
                Null-Ponter-Exception. Nun wird dieser Befehl ignoriert und es wird in den Log
                ausfuehrliche Fehlermeldungen gechrieben.


Version 01.05.00

    26.10.2021  Die SolvisMax6 realisiert offenbar den Startzaehler fuer den Brenner intern nur als
                vorzeichenlosen 16-Bit-Zaehler. Dadurch springt der Zaehler nach 65535 auf 0. Um
                trotzdem den echten Wert im Smarthome-System zu erhalten, wurde die base.xml noch
                um den eine Additions- und Multiplikations-Moeglichkeit der ermittelten Werte
                erweitert. Die Defintion der Korrekturwerte erfolgt nun in der
                "ModifiedChannelValues"-Sektion. Die alte "FixChannelValues"-Sektion wurde dadurch
                ueberfluessig.
                Wenn ein Ueberlauf beobachtet wird, muss entsprechend die base.xml angepasst
                werden.

    02.11.2021  Der Befehl SERVICE_RESET setzt nun auch einen normalen User-Access zurueck.
    
    19.11.2021  Die Definition der Freigabezeit der Messungen nach einem PowerOn erfolgt nun in der
                base.xml. Dazu wurde das Element "ModifiedChannelValues" in ChannelOptions
                umbenannt und die Delay-Zeit hinzugefuegt, um sie so allgemeiner zu halten.
                
    29.11.1992  ChannelOptions nochmal etwas umdefiniert, nun Korrekturfaktor und Offset gleichzeit
                moeglich.
                

src/de/sgollmer/solvismax/Constants.java
/************************************************************************
 * 
 * $Id$
 *
 * 
 ************************************************************************/

package de.sgollmer.solvismax;

import java.lang.reflect.Field;
import java.util.regex.Pattern;

import de.sgollmer.solvismax.log.LogManager.ILogger;
import de.sgollmer.solvismax.objects.Coordinate;

public class Constants {
	/**
	 * In case of an error the connection to the client ist closed delayed in case
	 * of sending the cause tro the client
	 */

	public static final int DELAYED_CLOSING_TIME = 1000;
	public static final String LOG4J_CONFIG_FILE = "log4j2.xml";

	public static final String CRLF = "\r\n";

	public static final String TINY_LOG_CONFIG_FILE = "tinylog.properties";
	public static final String TINY_LOG_FILE_PROPERTY_PREFIX = "writer";
	public static final String TINY_LOG_FILE_PROPERTY_SUFFIX = ".file";

	public static final String SUN_JAVA_COMMAND = "sun.java.command";

	public static final int MIN_TOUCH_TIME = 100;
	public static final int MAX_WAIT_TIME_ON_STARTUP = 120000;

	public static final int SCREEN_IGNORED_FRAME_SIZE = 3;
	public static final int CRYPT_PREVIOUS = 233;
	/**
	 * Max. number of tries, to get to the necessarry screen
	 */
	public static final int FAIL_REPEATS = 3;
	public static final int SET_REPEATS = 3;
	public static final int PREPARATION_REPEATS = 3;
	public static final int MAX_GOTO_DEEPTH = 10;
	public static final int LEARNING_RETRIES = 10;
	public static final int COMMAND_TO_QUEUE_END_AFTER_N_FAILURES = 3;
	public static final int COMMAND_IGNORED_AFTER_N_FAILURES = 10;

	public static final int HOURLY_EQUIPMENT_WINDOW_READ_INTERVAL_FACTOR = 3;

	public static final int MAX_CONNECTIONS = 50;
	public static final int RETRY_STARTING_SERVER_TIME = 60000;
	public static final int CYCLE_TIME_WHERE_QUEUE_EMPTY = 500;
	public static final int WAIT_TIME_AFTER_IO_ERROR = 10000;
	public static final int WAIT_TIME_AFTER_MQTT_ERROR = 100;
	public static final int ALIVE_TIME = 120000;
	public static final int FORCE_UPDATE_AFTER_N_INTERVALS = 3;
	public static final int WAIT_TIME_AFTER_THROWABLE = 30000;
	/**
	 * Maximum of time, within the screen isn't read again, if no other screen is
	 * selected
	 */
	public static final int TIME_COMMAND_SCREEN_VALID = 5000;

	public static final int SYNC_TOLERANCE_PERCENT = 10;

	public static final int MAX_OUTSIDE_TIME = 120000;
	public static final int SCREEN_SAVER_WIDTH_INACCURACY = 5;
	public static final int SCREEN_SAVER_HEIGHT_INACCURACY = 5;

	public static final int WAIT_AFTER_FIRST_ASYNC_DETECTION = 100;

	public static final long SETTING_TIME_RANGE_LOWER = 5000;
	public static final long SETTING_TIME_RANGE_UPPER = 50000;
	public static final int TIME_ADJUSTMENT_PROPOSAL_FACTOR_PERCENT = 150;
	public static final int TIME_ADJUSTMENT_MS_N = 10 * 60000; // Possible time adjustment of the solvis time every full
																// 10
																// minutes
	public static final int TIME_FINE_ADJUSTMENT_MS_N = 5 * 60000; // Possible time adjustment of the solvis time every
																	// full 10
	// minutes
	public static final int AVERAGE_COUNT_SOLVIS_CLOCK_PRECISION_CALCULATION = 10;
	public static final double PRECISION_DOUBLE = 0.000001D;

	public static final int MAX_WAIT_TIME_TERMINATING_OTHER_SERVER = 60000;
	public static final int NUMBER_OF_CONTROL_FILE_DUPLICATES = 3;

	public static final int MODBUS_SLAVE_ID = 1;

	public static final String[] CRYPT_NOT_CONFIGURED_VALUES = new String[] { "none", "aes-coded" };
	public static final double CIRCLE_AREA_PRECISION = 0.2;

	public static class Commands {
		public static final int REHEATING_PRIORITY = 10;
		public static final int EQUIPMENT_ON_OFF_PRIORITY = 5;
	}

	public static class Solvis {
		public static final int INPUT_BUFFER_SIZE = 1024;
		public static final String DISPLAY = "/display.bmp?";
		public static final String XML = "/sc2_val.xml";
		public static final String TOUCH = "/Touch.CGI?";
		public static final String X = "x=";
		public static final String Y = "y=";
		public static final String TASTER = "/Taster.CGI?taste=";
		public static final String ID = "i=";
		public static final Coordinate RELEASE_COORDINATE = new Coordinate(260, 260);
		public static final int INTERRUPT_AFTER_N_TOUCHES = 5; // After 5 touches, the screen
																// command can be interrupted
	}

	public static class Defaults {
		public static final int REHEATING_NOT_REQUIRED_ACTIVE_TIME = 30000;

	}

	public static class Files {
		public static final int INPUT_BUFFER_SIZE = 8192;
		public static final String RESOURCE_DESTINATION = "SolvisServerData";
		public static final String LEARN_DESTINATION = "LearnedImages";
		public static final String APPENDIX_DESTINATION = "Appendix";
		public static final String SOLVIS_ERROR_DESTINATION = "SolvisErrorImages";
		public static final String RESOURCE = "data";
		public static final String ERROR_SCREEN_PREFIX = "ErrorScreen-";
		public static final String GRAFIC_SUFFIX = "png";

		public static final String SOLVIS_SCREEN = "SolvisScreen." + Files.GRAFIC_SUFFIX;

		public static final Pattern ERROR_SCREEN_REGEX = Pattern
				.compile(Files.ERROR_SCREEN_PREFIX + "\\d+?\\." + Files.GRAFIC_SUFFIX);
		public static final int MAX_NUMBER_OF_ERROR_SCREENS = 100;
		public static final String CSV_ALL_CHANNELS = "SolvisAllChannels.csv";
		public static final String CSV_DOCUMENTATION = "SolvisDocumentation.csv";
		public static final String ZIP_PREFIX = "Images";
	}

	public static class XmlStrings {
		public static final String XML_MEASUREMENT_BOOLEAN = "BooleanValue";
		public static final String XML_MEASUREMENT_INTEGER = "IntegerValue";
		public static final String XML_MEASUREMENT_STRING = "StringValue";
		public static final String XML_MEASUREMENT_MODE = "ModeValue";
	}

	public static class ExitCodes {
		public static final int OK = 0;
		public static final int READING_CONFIGURATION_FAIL = 10;
		public static final int SERVER_PORT_IN_USE = 11;
		public static final int SERVER_TERMINATION_FAIL = 12;
		public static final int RESTART_FAILURE = 13;
		public static final int LEARNING_FAILED = 14;
		public static final int LEARNING_NECESSARY = 15;
		public static final int ARGUMENT_FAIL = 16;
		public static final int CRYPTION_FAIL = 17;
		public static final int MAILING_ERROR = 18;
		public static final int BASE_XML_ERROR = 19;
		public static final int XML_VERIFICATION_ERROR = 20;
		public static final int MQTT_ERROR = 21;
		public static final int TASK_CREATING_ERROR = 22;
		public static final int SOLVIS_UNIT_ERROR = 23;
	}

	public static class Mqtt {
		public static final String[] CMND_SUFFIXES = new String[] { //
				"/+/server/" + Mqtt.CMND_SUFFIX, // Server commands
				"/+/" + Mqtt.ONLINE_STATUS, //
				"/+/+/server/" + Mqtt.CMND_SUFFIX, // Server commands
				"/+/+/+/" + Mqtt.CMND_SUFFIX, // SET
				"/+/+/+/" + Mqtt.UPDATE_SUFFIX, // GET
				// "/#"
		};
		public static final String DATA_SUFFIX = "data";
		public static final String META_SUFFIX = "meta";
		public static final String CMND_SUFFIX = "cmnd";
		public static final String SERVER = "server";
		public static final String SCREEN_PREFIX = "screen";
		public static final String DEBUG_CHANNEL_PREFIX = "debugChannel";
		public static final String ONLINE_STATUS = "online";
		public static final String UPDATE_SUFFIX = "update";
		public static final String STATUS = "status";
		public static final String HUMAN_ACCESS = "human_access";
		public static final String ERROR = "error";
		public static final int MAX_INFLIGHT = 200;
		public static final int MIN_CONNECTION_REPEAT_TIME = 1000;
		public static final int MAX_CONNECTION_REPEAT_TIME = 120000;
		public static final String CONTROL = "gui_access";
		public static final Pattern CHANNEL_NAME_FROM = Pattern.compile("([A-Z]\\d+):(.*)");
		public static final String[] EMPTY_ARRAY = new String[0];
	}

	public static class Debug {

		public static final int USER_ACCESS_TIME = 60000;
		public static final boolean NO_MAIL = false; // kein Mailversand
		public static final boolean SCREEN_SAVER_DETECTION = false; // more debugging info
		public static final boolean DEBUG_TWO_STATIONS = false;

		public static final boolean SOLVIS_RESULT_NULL = false;

		public static final boolean BURST = false;

		public static final boolean OVERWRITE_ONLY_ON_LEARN = false;

		private static NoDebug[] NO_DEBUGS = new NoDebug[] { new NoDebug(BaseData.class, "DEBUG", false),
				new NoDebug("NO_MAIL", false), new NoDebug("SCREEN_SAVER_DETECTION", false),
				new NoDebug("DEBUG_TWO_STATIONS", false), new NoDebug("SOLVIS_RESULT_NULL", false),
				new NoDebug("BURST", false), new NoDebug("OVERWRITE_ONLY_ON_LEARN", false) };

		private static class NoDebug {
			private final Class<?> klass;
			private final String varName;
			private final boolean value;

			public NoDebug(final Class<?> klass, final String varName, final boolean value) {
				this.klass = klass;
				this.varName = varName;
				this.value = value;
			}

			public NoDebug(final String varName, final boolean value) {
				this(Debug.class, varName, value);
			}
		}

		public static void logDebugging(final ILogger logger) {
			boolean debugging = false;
			;
			for (NoDebug noDebug : NO_DEBUGS) {
				try {
					Field field = noDebug.klass.getField(noDebug.varName);
					boolean value = field.getBoolean(null);
					if (value != noDebug.value) {
						if (!debugging) {
							logger.warn("Debugging is activated, enabled debugging:");
						}
						logger.warn("    " + noDebug.varName);
					}
				} catch (NoSuchFieldException | SecurityException | IllegalArgumentException
						| IllegalAccessException e) {
					logger.error("Reflection error on DEBUG class", e);
				}
			}

		}

	}

	public static class Csv {
		public static final String ID = "Id";
		public static final String NAME = "Name";
		public static final String UNIT = "Unit";
		public static final String WRITE = "Write";
		public static final String MQTT = "MQTT suffix";
		public static final String CHANNEL_TYPE = "Channel type";
		public static final String BUFFERED = "Buffered";
		public static final String GLITCH_INHIBIT = "Glitch inhibit";
		public static final String DIVISOR = "Divisor";
		public static final String AVERAGE = "Average";
		public static final String DELAY_AFTER_ON = "Delay after on";
		public static final String FAST = "Fast";
		public static final String STRATEGY = "Strategy";
		public static final String OPTIONAL = "Optional";
		public static final String INCREMENT = "Increment";
		public static final String LEAST = "Least";
		public static final String MOST = "Most";
		public static final String INCREMENT_CHANGE = "Increment change";
		public static final String CHANGED_INCREMENT = "Changed increment";
		public static final String MODES = "Modes";
		public static final String ALIAS = "Alias";
		public static final String[] HEADER = new String[] { ID, NAME, UNIT, WRITE, MQTT, CHANNEL_TYPE, BUFFERED,
				GLITCH_INHIBIT, DIVISOR, AVERAGE, DELAY_AFTER_ON, FAST, STRATEGY, OPTIONAL, INCREMENT, LEAST, MOST,
				INCREMENT_CHANGE, CHANGED_INCREMENT, MODES, ALIAS };
	}

	public static class IoBroker {
		public static final String PAIRING_SCRIPT_NAME = "solvis_pairing.js";
		public static final String OBJECT_LIST_NAME = "SolvisSmartHomeServer.json";
		public static final String DEFAULT_MQTT_INTERFACE = "mqtt-client.0";
		public static final String DEFAULT_JAVASCRIPT_INTERFACE = "javascript.0";
		public static final String DEFAULT_IOBROKER_ID = "IoBroker";
	}
}
